pkg/executor/containerexecutor/containerexecutor.go

```diff
@@ -0,0 +1,844 @@
+	UsernameSecret   *testkube.SecretRef
+	TokenSecret      *testkube.SecretRef
+	Variables        map[string]testkube.Variable
+}
+
+// Logs returns job logs stream channel using kubernetes api
+func (c *ContainerExecutor) Logs(id string) (out chan output.Output, err error) {
+	out = make(chan output.Output)
+	logs := make(chan []byte)
+
+	go func() {
+		defer func() {
+			c.log.Debug("closing ContainerExecutor.Logs out log")
+			close(out)
+		}()
+
+		if err := c.TailJobLogs(id, logs); err != nil {
+			out <- output.NewOutputError(err)
+			return
+		}
+
+		for l := range logs {
+			entry, err := output.GetLogEntry(l)
+			if err != nil {
+				out <- output.NewOutputError(err)
+				return
+			}
+			out <- entry
+		}
+	}()
+
+	return
+}
+
+// Execute starts new external test execution, reads data and returns ID
+// Execution is started asynchronously client can check later for results
+func (c *ContainerExecutor) Execute(execution *testkube.Execution, options client.ExecuteOptions) (testkube.ExecutionResult, error) {
+	result := testkube.NewRunningExecutionResult()
+
+	ctx := context.Background()
+	err := c.createJob(ctx, *execution, options)
+	if err != nil {
+		return result.Err(err), err
+	}
+
+	podsClient := c.clientSet.CoreV1().Pods(c.namespace)
+	pods, err := c.GetJobPods(podsClient, execution.Id, 1, 10)
+	if err != nil {
+		return result.Err(err), err
+	}
+
+	l := c.log.With("executionID", execution.Id, "type", "async")
+
+	for _, pod := range pods.Items {
+		if pod.Status.Phase != corev1.PodRunning && pod.Labels["job-name"] == execution.Id {
+			// async wait for complete status or error
+			go func(pod corev1.Pod) {
+				_, err := c.updateResultsFromPod(ctx, pod, l, execution, result)
+				if err != nil {
+					l.Errorw("update results from jobs pod error", "error", err)
+				}
+			}(pod)
+
+			return result, nil
+		}
+	}
+
+	l.Debugw("no pods was found", "totalPodsCount", len(pods.Items))
+
+	return testkube.NewRunningExecutionResult(), nil
+}
+
+// Execute starts new external test execution, reads data and returns ID
+// Execution is started synchronously client will be blocked
+func (c *ContainerExecutor) ExecuteSync(execution *testkube.Execution, options client.ExecuteOptions) (testkube.ExecutionResult, error) {
+	result := testkube.NewRunningExecutionResult()
+
+	ctx := context.Background()
+	err := c.createJob(ctx, *execution, options)
+	if err != nil {
+		return result.Err(err), err
+	}
+
+	podsClient := c.clientSet.CoreV1().Pods(c.namespace)
+	pods, err := c.GetJobPods(podsClient, execution.Id, 1, 10)
+	if err != nil {
+		return result.Err(err), err
+	}
+
+	l := c.log.With("executionID", execution.Id, "type", "sync")
+
+	// get job pod and
+	for _, pod := range pods.Items {
+		if pod.Status.Phase != corev1.PodRunning && pod.Labels["job-name"] == execution.Id {
+			return c.updateResultsFromPod(ctx, pod, l, execution, result)
+		}
+	}
+
+	l.Debugw("no pods was found", "totalPodsCount", len(pods.Items))
+
+	return result, nil
+}
+
```

Comment:
Should we generalize Execute and ExecuteSync?

Reply:
We already have Execute and ExecuteSync. Are they fully different or can be reused?